### ISSUE TITLE
sandboxed builds with many sandbox rules fail on Darwin

### ISSUE THREAD

**Issue by mroi** [NixOS/nix#2311]
I am using Nix on macOS and I have sandboxed builds enabled. When I build texlive derivations using the `texlive.combine` function, the build can fail with this message:

```
these derivations will be built:
  /nix/store/1sjfz02piyqgplvmw8xdxz8gfaka1brj-texlive-combined-2017.drv
building '/nix/store/1sjfz02piyqgplvmw8xdxz8gfaka1brj-texlive-combined-2017.drv'...
Assertion failed: (diff <= INSTR_JUMP_NE_MAX_LENGTH), function push_jne_instr, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/Sandbox/Sandbox-851.200.128.41.1/src/compiler/patterns/serialize.c, line 186.
builder for '/nix/store/1sjfz02piyqgplvmw8xdxz8gfaka1brj-texlive-combined-2017.drv' failed due to signal 6 (Abort trap: 6)
error: build of '/nix/store/1sjfz02piyqgplvmw8xdxz8gfaka1brj-texlive-combined-2017.drv' failed
```

My guess is that this is caused by the sandbox profile containing too many rules. Disabling sandboxing and restarting `nix-daemon` causes the build to succeed. This is also currently my workaround: disable sandboxing whenever something fails like this, then reenable.

A better solution would probably be to check the length of the sandbox profile (somewhere around [here in `build.cc`](https://github.com/NixOS/nix/blob/master/src/libstore/build.cc#L2909) I guess) and fall back to a simpler profile automatically. How do you guys feel about this? I am willing to help getting this solved.

**Comment by edolstra** [comment]
Ping @copumpkin, @LnL7.

**Comment by copumpkin** [comment]
Wondering if this would still happen if we stopped doing as much "codegen" and started moving towards a more "data + code" sandbox design, as I've talked about in the past. That is, we'd factor out the common sandbox patterns into scheme on the sandbox side, rather than C++ on the Nix side.

**Comment by lheckemann** [comment]
>Disabling sandboxing and restarting nix-daemon causes the build to succeed

As an easier workaround (but a workaround nonetheless) I think you should be able to pass `--no-sandbox` on the command line as long as you're root or listed in `trusted-users`.

**Comment by mroi** [comment]
Ah, did not know that. Thanks for the hint.

**Comment by angerman** [comment]
Just ran into this today. Could run `nix-build` with `--option sandbox true` and build it successfully nor with `--option sandbox false`, due to some cc tools confusion.

Ended up doing `--option sandbox true` and then `nix-build` those failing derivations by hand. Not idea. If anyone iwth some more insight has a TODO list of items to be done to get sandboxes working properly on macOS, let me know!

**Comment by adrian-gierakowski** [comment]
I'm experiencing this as well. Any idea what it would take to fix it (so that we don't have to disable sandboxing)?

**Comment by adrian-gierakowski** [comment]
This is still relevant

**Comment by domenkozar** [comment]
Same as https://github.com/NixOS/nix/issues/4119

**Comment by aviallon** [comment]
Still relevant for me on arch

**Comment by adrian-gierakowski** [comment]
Looks like this hasn’t been fixed yet

**Comment by mroi** [comment]
Yes, indeed. Thank you so much for this nice piece of work. I deployed your patch here and can build `texlive` without issues in the sandbox.
